### PR TITLE
OCPBUGS-58467: Add range check for --parallel-images and --parallel-l…

### PR DESCRIPTION
**Pull request by lmzuccarelli**
…ayers flags

# Description

Add range checks

Github / Jira issue:  OCPBUGS-48467

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code Improvements (Refactoring, Performance, CI upgrades, etc)
- [ ] Internal repo assets (diagrams / docs on github repo)
- [ ] This change requires a documentation update on openshift docs

# How Has This Been Tested?

Tested locally with flag --parallel-images 0 and 11 and --parallel-layers 0 and 11

## Expected Outcome
Should fail if out of range

### CODE REVIEW

File: v2/internal/pkg/cli/executor.go [hunk 1]
```diff
@@ -374,6 +374,13 @@ func (o ExecutorSchema) Validate(dest []string) error {
 			return fmt.Errorf("--since flag needs to be in format yyyy-MM-dd")
 		}
 	}
+	// OCPBUGS-58467
+	if o.Opts.ParallelImages > 10 || o.Opts.ParallelImages < 1 {
+		return fmt.Errorf("the flag parallel-images must be between the range 0 to 10")
+	}
```
Comment (MayXuQQ) [inline v2/internal/pkg/cli/executor.go:380]: range 1 to 10? 0 is not allowed

Comment (lmzuccarelli) [reply v2/internal/pkg/cli/executor.go:380]: Yes - also I figured 10 should be max as we have had complaints about oc-mirror v2 when default setting were 10 for the parallel-images and 8 for the parallel-layers

Comment (MayXuQQ) [reply v2/internal/pkg/cli/executor.go:380]: How about changing the message 0 to 1?

> the flag parallel-images must be between the range 0 to 10

Comment (MayXuQQ) [reply v2/internal/pkg/cli/executor.go:380]: Where is the code checking the parallel-layers value?

Comment (lmzuccarelli) [reply v2/internal/pkg/cli/executor.go:380]: @MayXuQQ - code for parallel-layers is just below :)

File: v2/internal/pkg/cli/executor.go [hunk 2]
```diff
@@ -187,7 +187,7 @@ func NewMirrorCmd(log clog.PluggableLoggerInterface) *cobra.Command {
 		Example:       mirrorExamples,
 		Args:          cobra.MinimumNArgs(1),
 		SilenceErrors: true,
-		SilenceUsage:  false,
+		SilenceUsage:  true,
```
Comment (r4f4) [inline v2/internal/pkg/cli/executor.go]: Why are you undoing this change? I changed it because of https://github.com/openshift/oc-mirror/pull/1200#pullrequestreview-2973920153
See https://github.com/openshift/oc-mirror/pull/1200/commits/815cfded57b54ea0da7de650d9e8881d68bce9a8 for the reason.

File: v2/internal/pkg/cli/executor_test.go [hunk 3]
```diff
@@ -504,7 +507,19 @@ func TestExecutorValidate(t *testing.T) {
...
+		// check ParallelImages
+		opts.ParallelImages = 11
+		err = ex.Validate([]string{"file://test"})
+		assert.Equal(t, "the flag parallel-images must be between the range 1 to 10", err.Error())
+
+		// check ParallelLayerImages
+		opts.ParallelImages = 5
+		opts.ParallelLayerImages = 11
+		err = ex.Validate([]string{"file://test"})
+		assert.Equal(t, "the flag parallel-layers must be between the range 1 to 10", err.Error())
```
Comment (r4f4) [inline v2/internal/pkg/cli/executor_test.go]: Semantically these don't belong to this test case. They should be in their own `"validate should fail"`. Also instead of `assert.Equal` let's use `assert.Error` which prints a more clear message when the test fails.

File: v2/internal/pkg/cli/executor.go (hunk 2 above)
Comment (lmzuccarelli) [reply v2/internal/pkg/cli/executor.go]: It shows the help layout on the console before the error. Horrible user experience

Comment (r4f4) [reply v2/internal/pkg/cli/executor.go]: How is that horrible? If you use the oc-mirror flags in the wrong way (or the wrong flags) the help output will display the correct flags and their description.

File: v2/internal/pkg/cli/executor_test.go [hunk 4]
```diff
@@ -564,6 +567,48 @@ func TestExecutorValidate(t *testing.T) {
...
+		// check ParallelImages
+		opts.ParallelImages = 11
+		err := ex.Validate([]string{"file://test"})
+		assert.Error(t, err)
+
+		// check ParallelLayerImages
+		opts.ParallelImages = 5
+		opts.ParallelLayerImages = 11
+		err = ex.Validate([]string{"file://test"})
+		assert.Error(t, err)
```
Comment (r4f4) [inline v2/internal/pkg/cli/executor_test.go]: Let's add tests for `0` and negative values as well so we test both conditions in the validation check.

File: v2/internal/pkg/cli/executor_test.go [hunk 5]
```diff
@@ -564,6 +567,48 @@ func TestExecutorValidate(t *testing.T) {
 		opts.Global.WorkingDir = "" // reset
 		assert.Equal(t, "when destination is docker://, either --from (assumes disk to mirror workflow) or --workspace (assumes mirror to mirror workflow) need to be provided", ex.Validate([]string{"docker://test"}).Error())
 	})
+
+	t.Run("Testing Executor : validate should fail", func(t *testing.T) {
```
Comment (r4f4) [inline v2/internal/pkg/cli/executor_test.go]: Since you're creating a "should fail" test case, let's use the opportunity to refactor the previous "should pass" testcase and move all the checks for failures into this testcase?

File: v2/internal/pkg/cli/executor_test.go [hunk 6]
```diff
@@ -499,70 +514,92 @@ func TestExecutorValidate(t *testing.T) {
...
 		}
 
+		// check ParallelImages
+		opts.ParallelImages = 11
 		err := ex.Validate([]string{"file://test"})
-		if err != nil {
-			t.Fatalf("should not fail")
-		}
+		expectedErr := "the flag parallel-images must be between the range 1 to 10"
+		assert.EqualErrorf(t, err, expectedErr, "Error should be: %s, got: %w", expectedErr, err)
```
Comment (r4f4) [inline v2/internal/pkg/cli/executor_test.go]: `EqualErrorf` doesn't understand the `%w` verb it's specific to `fmt.Errorf`:
```
--- FAIL: TestExecutorValidate (0.00s)
    --- FAIL: TestExecutorValidate/Testing_Executor_:_validate_should_fail (0.00s)
        executor_test.go:521: 
            	Error Trace:	~/prjs/github.com/openshift/oc-mirror/v2/internal/pkg/cli/executor_test.go:521
            	Error:      	Error message not equal:
            	           	expected: "the flag parallel-images must be between the range 1 to 10 banana"
            	           	actual  : "the flag parallel-images must be between the range 1 to 10"
            	Test:       	TestExecutorValidate/Testing_Executor_:_validate_should_fail
            	Messages:   	Error should be: the flag parallel-images must be between the range 1 to 10 banana, got: %!w(*errors.errorString=&{the flag parallel-images must be between the range 1 to 10})
```
Also if the assert message is just going to say the errors don't match, we don't need `EqualErrorf`, `EqualError` suffices:
```suggestion
		assert.EqualError(t, err,  "the flag parallel-images must be between the range 1 to 10")
```
```
--- FAIL: TestExecutorValidate (0.00s)
    --- FAIL: TestExecutorValidate/Testing_Executor_:_validate_should_fail (0.00s)
        executor_test.go:521: 
            	Error Trace:	~/prjs/github.com/openshift/oc-mirror/v2/internal/pkg/cli/executor_test.go:521
            	Error:      	Error message not equal:
            	           	expected: "the flag parallel-images must be between the range 1 to 10 banana"
            	           	actual  : "the flag parallel-images must be between the range 1 to 10"
            	Test:       	TestExecutorValidate/Testing_Executor_:_validate_should_fail
```

**Review by r4f4 — COMMENTED** [review]
/lgtm
